channel/summons.go

```diff
@@ -0,0 +1,203 @@
+package channel
+
+import (
+	"time"
+
+	"github.com/Hucaru/Valhalla/common/opcode"
+	"github.com/Hucaru/Valhalla/mpacket"
+)
+
+// Summon represents a player-owned summon/puppet instance.
+type Summon struct {
+	OwnerID int32
+	SkillID int32
+	Level   byte
+	HP      int
+
+	Pos       pos
+	Stance    byte
+	Foothold  int16
+	ExpiresAt time.Time
+
+	// Flags
+	IsPuppet   bool
+	SummonType int32
+}
+
+type summonState struct {
+	puppet  *Summon
+	summon  *Summon
+	puTimer *time.Timer
+	smTimer *time.Timer
+}
+
+func (p *player) ensureSummonState() {
+	if p.summons == nil {
+		p.summons = &summonState{}
+	}
+}
+
+// addSummon registers and broadcasts a summon; replaces any existing same-type summon.
+func (s *Server) addSummon(p *player, su *Summon, durationSec int) {
```

Comment:
I don't think the server needs to keep track of the summons. They should be attached to the player. When summon information needs to be broadcast at most it's only ever to the player's instance.